policy: reduce allocs by avoiding use of interface types, retire MapStateOwners

jrajahalme opened the pull request:
Reduce allocations in policy map computation by:

- Change the `container/bitlpm.Trie` type to use `Key[K]` as a type constraint instead of the stored key type. This allows the actual key type to be stored by value, reducing needed memory allocations and indirection.

- Remove the Iterator interface type and return the iterators by value so that they can be stored in the caller's stack instead of in the heap. This change makes the use of the explicit iterators faster than the callback based alternatives.

- Remove the tracking of `MapStateOwners` made redundant by the change of FQDN identities to not mutate their labels [(#32769)](https://github.com/cilium/cilium/pull/32769)
  - Mutation of `reserved:host` identity is still possible and is handled by not issuing incremental deletion for the change and informing the caller (ipcache) to trigger forced policy regenerations in this case.

- Preallocate policy MapState map using the size of the previous policy map.

These changes speed up the CIDR deny policy benchmark
(`BenchmarkRegenerateCIDRDenyPolicyRules`) by ~36%, reduces the memory
needed by ~67%, and reduces the number of heap allocations by ~85%.

These changes bring the `BenchmarkRegenerateCIDRDenyPolicyRules` in v1.17 to be ~10000x faster than in v1.16, with 99,5% reduction in memory consumption (99,6% reduction in the number of heap allocations).

`BenchmarkRegenerateCIDRPolicyRules` (without the deny rules) is now ~3.6x faster than v1.16, with 80% reduction in memory consumption (99% reduction in the number of heap allocations).

jrajahalme committed:
bitlpm: Remove the Iterator interface

Remove the Iterator interface and return the implementation of the
iterator directly by value. This removes a heap allocation for each
iteration, as the iterator is allocated in the stack of the calling
function.

This change speeds up the CIDR deny policy benchmark
(BenchmarkRegenerateCIDRDenyPolicyRules) by ~12%, reduces the memory
needed by ~13%, and reduces the number of heap allocations by ~65%.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
bitlpm: Use Key[K] as a type constraint

Use Key[K] as a type constraint, rather than the type of the stored key,
or the key parameter. With this change the generic type K within the type
and function parameters is the actual key type, rather than an interface,
avoiding indirection and taking less space.

Nil checks can no longer performed on the key parameter as it is no
longer an interface. If 'K' is a pointer type, its implementations of
Key[K] interface functions must accept nil receiver and perform nil
checks itself. If 'K' is small it should be passed by value for better
performance as this avoids memory indirections.

This change alone speeds up the CIDR deny policy benchmark
(BenchmarkRegenerateCIDRDenyPolicyRules): by ~12%, reduces the memory
needed by ~10%, and reduces the number of heap allocations by ~50%.

This change is also needed for removal of the Iterator interface in a
later commit.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Simplify mapState.upsert

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Fix comments

Fix mispelling of 'mapStateEntry' and commenta in LPMAncestors().

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Remove mapstate owner tracking

The set of MapStateEntry owners was added to keep track of which
selectors need a specific MapStateEntry with a specific identity to be
present, so that when an identity is incrementally deleted from one
selector, the MapStateEntry is actually deleted when the identity is
removed from all the selectors that had it. Now that selector cache is
transactional we can rely on any deleted identity to be removed from all
selectors within the same incremental map change. Thus we no longer need
to track the MapStateOwners.

This change speeds up the CIDR deny policy benchmark
(BenchmarkRegenerateCIDRDenyPolicyRules) by ~10%, reduces the memory
needed by ~25%, and reduces the number of heap allocations by ~8%.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Rename newMapState() as emptyMapState()

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

nathanjsweet reviewed (approved):
LGTM

jrajahalme committed:
policy: Remove unused parameter from test code

Remove unused 'epLabels' parameter from 'policyDistillery.distillPolicy()'.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Add unit test for CIDR & FQDN incremental identity deletion

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
policy: Prealloc mapstate using previous size

Reduce allocations and memory use by preallocating the mapstate map with
the capacity of the size of the previous map. This minimizes allocations
for policy updates, but also means that the policy map shrinks only after
the 2nd policy recompute.

This change speeds up the CIDR deny policy benchmark
(BenchmarkRegenerateCIDRDenyPolicyRules) by ~9%, reduces the memory
needed by ~44%, and reduces the number of heap allocations by ~4%.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

jrajahalme committed:
labels: Remove debug printfs

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>

squeed commented on pkg/container/bitlpm/trie.go:
```diff
@@ -101,16 +101,6 @@ type Trie[K Key[K], T any] interface {
 	ForEach(fn func(uint, K, T) bool)
 }
 
-// Iterator is an interface that can be used to produce the next key/value pair in iteration
```
I generally don't prefer  returning unexported types, because it can lead to awkwardness where variable types cannot be spelled. Thus, one must always use `:=`.

```go
i := trie.AncestorIterator()
```

For example, this code becomes impossible:
```go
var i trie.Iterator 
if foo {
  i = trie.AncestorIterator()
} else {
  i = trie.AncestorLongestPrefixFirstIterator()
}
...
```

However, the stack allocation is useful. Maybe keep the Iterator interface for convenience?

squeed replied:
And maybe we should add a TODO to turn these in to rangefuncs?

squeed replied:
(feel free to self-resolve)

jrajahalme replied:
Nested range funcs have horrible performance (https://github.com/golang/go/issues/69015). These iterators are used from rangefuncs and this combination has the best performance and the business logic using the (outer) rangefunc iterators is very concise.

jrajahalme replied:
We can add the Iterator interface back if/when needed. Having it might entice using it which would lead to worse performance.

jrajahalme committed:
policy: Block incremental deletion of mutated identities

The host identity can be mutated, e.g., when the kube-apiserver is
removed from the local host. When this happens, the
'reserved:kube-apiserver' label is removed from the host identity (1),
causing any selectors selecting the 'reserved:kube-apiserver' label to no
longer select the host identity (1). Since we no longer have any
reference counting (such as owner tracking) in the policy map state, we
must block the incremental deletion of any such mutated identities to keep
traffic to/from that identity allowed in case some other selector may be
still allowing such traffic.

For this strategy to work, the entity mutating the host identity must
also trigger forced policy regenerations for all endpoints so that the
policy map entries with the host identity are removed if actually not
required due to any other selectors.

To implement this in a generic was we add a "mutated" boolean return
value to SelectorCache.UpdateIdentities function. This function is
already explicitly aware of such mutated identities. Instead of informing
the SelectorCache user's of a deleted identity in this case, we simply
just update the cached selector that no longer selects the mutated
identity and return "mutated = true" to inform the caller to trigger
forced policy updates on all endpoints when this happens.

IPCache metadata logic is changed to make note of the new "mutated"
return value and trigger policy updates when 'true' is returned.

Signed-off-by: Jarno Rajahalme <jarno@isovalent.com>
